File: src/lib/auth/jwt.ts
// JWT token generation and verification for OAA Auth System
// Provides stateless session tokens with configurable expiry

import { generateHMAC, generateToken } from '../crypto/hash';

const JWT_SECRET = process.env.JWT_SECRET || 'oaa-dev-secret-change-in-production';
const JWT_EXPIRES_IN_DAYS = 30;
const JWT_EXPIRES_IN_MS = JWT_EXPIRES_IN_DAYS * 24 * 60 * 60 * 1000;

export interface TokenPayload {
  userId: string;
  handle: string;
  email?: string;
  walletAddress?: string;
  iat?: number; // Issued at
  exp?: number; // Expires at
}

/**
 * Base64URL encode (JWT-safe)
 */
function base64UrlEncode(data: string): string {
  return Buffer.from(data)
    .toString('base64')
    .replace(/\+/g, '-')
    .replace(/\//g, '_')
    .replace(/=/g, '');
}

/**
 * Base64URL decode
 */
function base64UrlDecode(data: string): string {
  const padded = data + '==='.slice(0, (4 - (data.length % 4)) % 4);
  return Buffer.from(
    padded.replace(/-/g, '+').replace(/_/g, '/'),
    'base64'
  ).toString();
}

/**
 * Generate JWT access token
 */
export function generateJWT(payload: Omit<TokenPayload, 'iat' | 'exp'>): string {
  const now = Date.now();
  const fullPayload: TokenPayload = {
    ...payload,
    iat: now,
    exp: now + JWT_EXPIRES_IN_MS,
  };

  const header = base64UrlEncode(JSON.stringify({ alg: 'HS256', typ: 'JWT' }));
  const body = base64UrlEncode(JSON.stringify(fullPayload));
  const signature = base64UrlEncode(generateHMAC(`${header}.${body}`, JWT_SECRET));

  return `${header}.${body}.${signature}`;
}

/**
 * Verify and decode JWT token
 * Returns null if invalid or expired
 */
export function verifyJWT(token: string): TokenPayload | null {
  try {
    const parts = token.split('.');
    if (parts.length !== 3) return null;

    const [header, body, signature] = parts;

    // Verify signature
    const expectedSignature = base64UrlEncode(
      generateHMAC(`${header}.${body}`, JWT_SECRET)
    );
    if (signature !== expectedSignature) return null;

    // Decode payload
    const payload: TokenPayload = JSON.parse(base64UrlDecode(body));

    // Check expiration
    if (payload.exp && payload.exp < Date.now()) return null;

    return payload;
  } catch (error) {
    return null;
  }
}

/**
 * Decode token without verification (for debugging only)
 */
export function decodeJWT(token: string): TokenPayload | null {
  try {
    const parts = token.split('.');
    if (parts.length !== 3) return null;

    return JSON.parse(base64UrlDecode(parts[1]));
  } catch (error) {
    return null;
  }
}

/**
 * Extract token from Authorization header
 */
export function extractBearerToken(authHeader?: string | null): string | null {
  if (!authHeader || !authHeader.startsWith('Bearer ')) {
    return null;
  }
  return authHeader.substring(7);
}

/**
 * Check if token is expired or about to expire
 */
export function isTokenExpiring(token: string, bufferMs: number = 60000): boolean {
  const payload = verifyJWT(token);
  if (!payload || !payload.exp) return true;
  return payload.exp - Date.now() < bufferMs;
}

/**
 * Generate a refresh token (longer-lived, stored in DB)
 */
export function generateRefreshToken(): string {
  return generateToken(48); // 384 bits of entropy
}

File: app/auth/jwt.py
# app/auth/jwt.py
"""
JWT Token Verification for OAA Auth System

Compatible with the TypeScript JWT implementation in src/lib/auth/jwt.ts
Uses HMAC-SHA256 signature verification.
"""

import os
import hmac
import hashlib
import base64
import json
import time
from dataclasses import dataclass
from typing import Optional

# JWT secret - must match the TypeScript JWT_SECRET
JWT_SECRET = os.getenv("JWT_SECRET", "oaa-dev-secret-change-in-production")

if JWT_SECRET == "oaa-dev-secret-change-in-production":
    import warnings
    warnings.warn("JWT_SECRET is using default value. Set JWT_SECRET in production!")


@dataclass
class AuthClaims:
    """Authenticated user claims from JWT"""
    user_id: str  # 'sub' field in JWT (subject_id)
    handle: str
    email: Optional[str] = None
    wallet_address: Optional[str] = None
    iat: Optional[int] = None  # Issued at timestamp (ms)
    exp: Optional[int] = None  # Expiration timestamp (ms)


def base64url_encode(data: bytes) -> str:
    """Base64URL encode (JWT-safe, no padding)"""
    return base64.urlsafe_b64encode(data).rstrip(b'=').decode('utf-8')


def base64url_decode(data: str) -> bytes:
    """Base64URL decode with padding handling"""
    # Add padding if needed
    padding = 4 - (len(data) % 4)
    if padding != 4:
        data += '=' * padding
    return base64.urlsafe_b64decode(data)


def generate_hmac(data: str, secret: str) -> str:
    """Generate HMAC-SHA256 signature (hex)"""
    return hmac.new(
        secret.encode('utf-8'),
        data.encode('utf-8'),
        hashlib.sha256
    ).hexdigest()


def verify_jwt(token: str) -> Optional[AuthClaims]:
    """
    Verify and decode JWT token.
    
    Compatible with TypeScript JWT from src/lib/auth/jwt.ts
    Returns AuthClaims if valid, None if invalid or expired.
    
    Args:
        token: JWT string in format "header.payload.signature"
        
    Returns:
        AuthClaims if token is valid, None otherwise
    """
    try:
        parts = token.split('.')
        if len(parts) != 3:
            return None
        
        header_b64, payload_b64, signature_b64 = parts
        
        # Verify signature
        expected_sig = base64url_encode(
            bytes.fromhex(generate_hmac(f"{header_b64}.{payload_b64}", JWT_SECRET))
        )
        
        if signature_b64 != expected_sig:
            return None
        
        # Decode payload
        payload_json = base64url_decode(payload_b64).decode('utf-8')
        payload = json.loads(payload_json)
        
        # Check expiration (exp is in milliseconds)
        exp = payload.get('exp')
        if exp and exp < int(time.time() * 1000):
            return None
        
        # Extract claims
        return AuthClaims(
            user_id=payload.get('userId', ''),
            handle=payload.get('handle', ''),
            email=payload.get('email'),
            wallet_address=payload.get('walletAddress'),
            iat=payload.get('iat'),
            exp=exp
        )
        
    except Exception as e:
        # Log for debugging but don't expose details
        import sys
        print(f"JWT verification failed: {e}", file=sys.stderr)
        return None


def decode_jwt_unsafe(token: str) -> Optional[dict]:
    """
    Decode JWT without verification (for debugging only).
    
    WARNING: This does NOT verify the signature!
    Never trust data from this function for authentication.
    """
    try:
        parts = token.split('.')
        if len(parts) != 3:
            return None
        
        payload_json = base64url_decode(parts[1]).decode('utf-8')
        return json.loads(payload_json)
    except Exception:
        return None


def is_token_expired(token: str) -> bool:
    """Check if token is expired without full verification"""
    payload = decode_jwt_unsafe(token)
    if not payload:
        return True
    
    exp = payload.get('exp')
    if not exp:
        return True
    
    return exp < int(time.time() * 1000)


def extract_bearer_token(auth_header: Optional[str]) -> Optional[str]:
    """
    Extract JWT from Authorization header.
    
    Args:
        auth_header: Authorization header value (e.g., "Bearer xyz...")
        
    Returns:
        Token string or None if not present/invalid format
    """
    if not auth_header or not auth_header.startswith('Bearer '):
        return None
    return auth_header[7:]  # Remove "Bearer " prefix
